### DISCUSSION TITLE
Testing with UPF mode enabled

### DISCUSSION THREAD

**Discussion by jovans2** [vhive-serverless/vHive#398] [Q&A] [answered]
Hello,

When I follow the instructions from https://github.com/ease-lab/vhive/blob/main/docs/profiling.md and set flags (UPFEnabled, SnapshotEnabled, MetricsMode) I run into following issue:

- on first invocation I get the error: Task could not be killed.
- on subsequent invocations: "Tuntap IOCTL TUNSETIFF failed [0], errno device or resource busy", and "Bridge could not be created"

In general, I cannot run any of the tests with set UPF flag and I cannot come to the point where UPF statistics are written into file.

Also, when running all tests on a single node cluster (with and without UPF flag) I get the warning that kn8 could not be reached and default dns is used instead. Does it affect my experiments?

Do you know how could I solve this problem?
Thank you,
Jovan

**Comment by ustiugov** [comment, marked as answer]
hi @jovans2 

it seems that your previous runs left some VMs running or some tap devices were not torn down. This usually happens when something crashed during the previous runs. You can try to use the cleaning script to clean all the resources: 
```bash
./scripts/clean_fcctr.sh
```

The warning regarding DNS is not an issue.

In general, when reporting bugs, please include log snippets in your description and attach full logs as files if you think it's necessary. Also, please specify the host specs (HW and SW).

**Reply by jovans2** [reply]
Hello @ustiugov,

I use CloudLab profile suggested in https://github.com/ease-lab/vhive/blob/main/docs/quickstart_guide.md (Utah). So the full sequence of my actions is:

1) git clone --depth=1 https://github.com/ease-lab/vhive.git
2) cd vhive
3) mkdir -p /tmp/vhive-logs
4) ./scripts/cloudlab/setup_node.sh
5) sudo screen -dmS containerd containerd
6) sleep 5
7) sudo PATH=$PATH screen -dmS firecracker /usr/local/bin/firecracker-containerd --config /etc/firecracker-containerd/config.toml
8) sleep 5
9) remove t.Skip("Skipping TestProfileSingleConfiguration") from perf_bench_test.go
10)   update flags to be true in vhive_test.go
        isUPFEnabledTest       = flag.Bool("upfTest", true, "Enable user-level page faults guest memory management")
	isSnapshotsEnabledTest = flag.Bool("snapshotsTest", true, "Use VM snapshots when adding function instances")
	isMetricsModeTest      = flag.Bool("metricsTest", true, "Calculate UPF metrics") 
11) source /etc/profile && go build
12) ./scripts/install_pmutool.sh
13)  sudo env "PATH=$PATH" go test -v -timeout 99999s -run TestProfileSingleConfiguration -args -funcNames helloworld -vm 1 -rps 20 -l 1

However I get the following error:

--- PASS: TestProfileSingleConfiguration (434.37s)
PASS
INFO[2021-09-08T10:10:50.555516468-06:00] waiting for goroutines                       
ERRO[2021-09-08T10:10:50.556500746-06:00] Failed to kill the task                       error="ttrpc: closed: unknown" vmID=0-0
WARN[2021-09-08T10:10:50.556552512-06:00] ttrpc: closed: unknown                        vmID=0-0
INFO[2021-09-08T10:10:50.556579836-06:00] waiting done                                 
INFO[2021-09-08T10:10:50.556593958-06:00] Closing fcClient                             
INFO[2021-09-08T10:10:50.556612778-06:00] Closing containerd client                    
INFO[2021-09-08T10:10:50.556638565-06:00] Removing bridges 

If I run without step 10 (so no flags) everything completes regularly.

Do you know how should I proceed? 
Thank you!

**Reply by ustiugov** [reply]
it appears to me that these are not significant errors but the peculiarity of tearing down VMs after taking a snapshot. Inspecting the code of the benchmark, I have to say that the benchmark only supports boot-based started properly but it can be easily extended. 

However, what exactly are you studying? there are several benchmarks so we can see which one suits your goals the best. the benchmark that you are looking at right now studies the implications of functions co-location on the same CPU. Is that what you'd like to study?

**Reply by jovans2** [reply]
@ustiugov 
I want to conduct an experiment to create a trace of all page faults per function invocation, get TLB hit/miss rate and other potential micro-architectural details (ideally for an invoker that has a distribution of function invocation).
Also, is it possible to use perf instead of toplev, since it is not intuitive what all those percentages mean? 
What would you suggest?

**Reply by ustiugov** [reply]
I suggest using taking this [test](https://github.com/ease-lab/vhive/blob/main/bench_test.go#L191) as a baseline. it allows to run function instances of different type and with a various number of parallel invocations, also it supports both vanilla snapshots and REAP (UPF) and generate metrics. 

as for profiling it, you can add `perf` runs to that benchmark, programmatically or externally, in the way you like it to work (e.g., particular counters). toplev uses perf internally and it's open-source so technically you can peek into what it does internally. However, I recommend using perf for your case.

As for the page fault tracing, you can peek into the trace file generated in the snapshot folder `/fccd/snapshots/<VM_ID>` but comment this [line](https://github.com/ease-lab/vhive/blob/899f98ee62cc9d2dcf08fcb630ead4f8b9fd3430/ctriface/bench_test.go#L94) because it removes the snapshot files after the benchmark finishes.

let me know if you have any other questions.

**Reply by jovans2** [reply]
Thank you for the answer, I will try to integrate perf tool and use it instead toplev. Regarding snapshots - I have them in /fccd/snapshots folder, but I do not know how to read those files (mem_file, working_set_pages, snap_file) they are not in the supported format?

**Reply by ustiugov** [reply]
what you need is the working_set_pages. these are guest-physical memory addresses in the order of accesses happening

**Reply by jovans2** [reply]
I understand that of course, but the file format is not readable, screenshot attached.

![ws_file](https://user-images.githubusercontent.com/69543158/132589743-117efd01-eb8b-4352-aa89-f04107ef67dd.png)

**Reply by ustiugov** [reply]
@jovans2 sorry, it's another file, called *trace*. It is only generated when ran with UPF enabled.

**Reply by jovans2** [reply]
I removed line (https://github.com/ease-lab/vhive/blob/899f98ee62cc9d2dcf08fcb630ead4f8b9fd3430/ctriface/bench_test.go#L94) but I cannot see "trace" file, it seems that it is still being erased. I can see files mem_file, working_set_pages, snap_file, but not "trace" file. 

Also, I removed the same line from vhive_test.go

**Reply by jovans2** [reply]
Solved the problem - WriteTrace was never invoked. Now things work fine, thank you!

**Reply by ustiugov** [reply]
nice! let me know if you have any more questions!

**Reply by jovans2** [reply]
Hello @ustiugov,

I am sorry for bothering you again - I managed to add PERF instead of TOPLEV for gathering metrics. However, it is not clear to me how can I ensure that PERF is measuring HW counters for the core where the computation is actually happening? Do you know how can I enforce that? 

Thank you!

**Reply by ustiugov** [reply]
hi @jovans2, you can bind a VM to a particular core, as we do it [here](https://github.com/ease-lab/vhive/blob/899f98ee62cc9d2dcf08fcb630ead4f8b9fd3430/perf_bench_test.go#L659)

**Reply by jovans2** [reply]
Hello @ustiugov,

Another question - to invoke many functions at the same time without waiting on the previous request to be completed - is it okay to use TestBenchParallelServe? If yes, do you know what is the upper bound of parallel invocations? The test fails whenever I set more than 50 parallel invocations. 
The other thing, do I need separate VM for every invocation (which seems reasonable for me - every call is a different user and we need to provide isolation) or invocations can share VM? 

Paper says "vHive follows the model of AWS Lambda, which deploys a single function inside a MicroVM that processes a single invocation at a time", is the call "resp, met, err := funcPool.Serve(context.Background(), vmIDString, imageName, "replay")" basically creation of a MicroVM or Pod?

Thank you!

**Reply by ustiugov** [reply]
hi @jovans2 

1. feel free to modify the test as you feel the need for.
2. the upper bound depends on the number of concurrently booting.loading instances because this is a relatively lengthy and resource-consuming procedure, which often results in timeouts. If the traffic is steered to warm VMs, one can push a lot more requests.
3. No, you don't need a separate instance for each request although it's common to maintain strict FIFO order of requests, i.e., one at a time, steered to each instance (e.g., in AWS Lambda).
4. yes, vHive by default supports 1 function instance deployed in one MicroVM. However, you are free to steer many requests to the gRPC server inside. Basically, it depends on your benchmark's code. You can configure it in the way you like.

**Reply by jovans2** [reply]
Thank you for the clarification. So I can start an instance, keep it warm and send as many requests as I want? Also, is there a way to implement or use grace period?

**Reply by ustiugov** [reply]
if you skim the code of the benchmark, it does all of what you mentioned. if you don't shut a VM down explicitly (with RemoveInstance), it would stay and it can keep serving as long as you like.
